supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,462 @@
+        method: str | None = None,
+        target_options: dict[str, Any] | None = None,
+    ) -> None:
+        """Run the benchmarking experiment and analyse the results.
+
+        Args:
+            num_circuits: Number of circuits to run.
+            layers: An iterable of the different layer depths to use during the experiment.
+            target: Either a local :class:`~cirq.SimulatorBase` to use or the name of a Superstaq
+                target. If None then a local simulator is used. Defaults to None.
+            shots: The number of shots to sample. Defaults to 10,000.
+            method: Optional method to use on the Superstaq device. Defaults to None corresponding
+                to normal running.
+            target_options: Optional configuration dictionary passed when submitting the job.
+        """
+        if self._samples is not None:
+            warnings.warn("Existing results will be overwritten.")
+
+        if any(depth <= 0 for depth in layers):
+            raise ValueError("The `layers` iterator can only include positive values.")
+
+        self._samples = self.build_circuits(num_circuits, layers)
+        self._clean_circuits()
+
+        if target is None:
+            target = cirq.Simulator()
+
+        if isinstance(target, str):
+            self.submit_ss_jobs(target, shots, method, **(target_options or {}))
+        else:
+            self.sample_circuits_with_simulator(target, shots)
+
+    def _clean_circuits(self) -> None:
+        """Removes any terminal measurements that have been added to the circuit and replaces
+        them with a single measurement of the whole system in the computational basis
+        """
+        for sample in self.samples:
+            if sample.circuit.has_measurements():
+                sample.circuit = cirq.drop_terminal_measurements(sample.circuit)
+            # Add measurement of qubit state in computational basis.
+            sample.circuit += cirq.measure(sample.circuit.all_qubits())
+
+    def submit_ss_jobs(
+        self,
+        target: str,
+        shots: int = 10_000,
+        method: str | None = None,
+        **target_options: Any,
+    ) -> None:
+        """Submit the circuit samples to the desired target device and store the resulting
+        probabilities.
+
+        The set of circuits is partitioned as necessary to not exceed the maximum circuits that can
+        be submitted to the given target device. The function then waits for the jobs to complete
+        before saving the resulting probability distributions.
+
+        Args:
+            target: The name of the target device.
+            shots: The number of shots to use. Defaults to 10,000
+            method: Optional method to use on the Superstaq device. Defaults to None corresponding
+                to normal running.
+            target_options: Optional configuration kwargs passed when submitting the job.
+        """
+        for sample in tqdm(self.samples):
```

Comment:
I couldn't work out a way to do this where I could also store data about each circuit within a job that contained multiple circuits. This is a feature that is quite important when doing the data analysis for the implemented experiments.

Reply:
> store data about each circuit within a job

sorry im probably missing something obvious but not sure i understand - could you point me to where this is happening?

Reply:
> the only problem is that each sub-job starts fresh, so we lose some efficiency in terms of api calls. it seems like it would be really nice if we added some methods to facilitate functionality into the job itself (ideally such that it automatically populates their job info with whatever we've already retrieved for the parent job to avoid extra work, or maybe even with a reference to the parent job so that some api calls only have to happen once for the batch? this second version might be a bit more messy than it's worth though)

Yeah, I've was thinking about this over lunch today. I might have a play tomorrow and see if I can come up with some kind of "BatchJob" object that facilitates this. I think that would be the most elegant solution and could potentially be really nice to have more widely. Might take a bit of work though (and would certainly need to be a different ticket)